docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
+* all values are propagated to a role through `r_` prefixed variables only
+* all values that are set inside a role are read outside of a role through `r_` prefixed variables only
+* all available `r_` prefixed variables are documented and checked before a main role task is run
+* a table of `r_` prefixed variables in README.md is automatically generated
+
+### Terminilogy
+
+**Role variables**: All `r_` prefixed variables read/set inside a role
+
+**Local variables**: All `l_` prefixed variables set inside a role
+
+### Role properties
+
+**Solid API**: all variables that a role caller can set need to be well defined.
+It means each such variable must have the following information attached:
+* name
+* description
+* default value (if the variable is defined)
+* set of acceptable values
+* if it is required or if it is optional
+
+Well defined API allows automatically generated documentation and variable checks.
+
+**Free of outer variables**: only variables consisting the role API or defined locally are allowed.
```

Comment:
I'm against using set_fact whenever possible.

I also don't care this pattern:
`
- include_role:
    name: openshift_node
  vars:
    r_openshift_node_kubelet_args: "{{ openshift_node_kubelet_args | default({}) }}"
`
The reason is because it's obscuring the role's variables.  If someone wants to investigate a role's behavior, they have to look all over our plays to figure out what's actually being passed into the role.

IMO, the only solution to these problems is to define sensible defaults and have values set in inventory/group_vars.  Why inventory/group_vars?  Because
1) We shouldn't have to rely on a specific play to setup the role 'just right' in order for the role to do it's work.  The role needs to be as independent of the play calling it as possible, plays for the most part should just 'call the role.'
2) Since we're providing plays to the user, plays are not user-modifiable.  If we were only providing a few roles, then users would create their own plays and can define variables however they want.  Since plays are not user-modifiable, we need to provide a pattern where users can easily modify the behavior of roles and plays, and this is done through inventory/group/host_vars/extra_vars.

I prefer inventory group and host_vars because they are implicitly tied to inventory, and I think it better fits the pattern of how ansible should be consumed.